File: ts/utils/autocmd.ts
declare var autocmdFunctions: { [id: string]: () => void }
autocmdFunctions = {}

declare var _G: any
_G["callAutoCmdFunction"] = (id: string) => {
  autocmdFunctions[id]()
}

type AutoCmdArgs = {
  id: string,
  events: string | string[],
  filetypes: string | string[],
  command: string | (() => void)
}
export const autocmd: (args: AutoCmdArgs) => void = ({ id, events, filetypes, command }) => {
  const commandText =
    typeof command === "string"
      ? command
      : `lua callAutoCmdFunction("${id}")`
  if (typeof command !== "string") {
    autocmdFunctions[id] = command
  }

  const eventsText =
    typeof events === "string"
      ? events
      : events.join(",")
  const fileTypesText =
    typeof filetypes === "string"
      ? filetypes
      : filetypes.join(",")

  vim.cmd("augroup myautocmd")
  vim.cmd(`autocmd! ${eventsText} ${fileTypesText} ${commandText}`)
  vim.cmd("augroup END")
}

File: lua/utils/autocmd.lua
autocmdFunctions = {}
_G.callAutoCmdFunction = function(id)
	autocmdFunctions[id]()
end

--@param id string
--@param events string[] | string
--@param filetypes string[] | string
--@param command string | function
return function(args)
	local id = args.id
	local command = args.command
	local events = args.events
	local filetypes = args.filetypes
	local commandText = ((type(command) == "string") and command) or (('lua callAutoCmdFunction("' .. id) .. '")')
	if type(command) ~= "string" then
		autocmdFunctions[id] = command
	end
	local eventsText = ((type(events) == "string") and events) or table.concat(events, ",")
	local fileTypesText = ((type(filetypes) == "string") and filetypes) or table.concat(filetypes, ",")
	vim.cmd("augroup " .. tostring(id))
	vim.cmd("autocmd!")
	vim.cmd((((("autocmd " .. eventsText) .. " ") .. fileTypesText) .. " ") .. commandText)
	vim.cmd("augroup END")
end
